actions/rules/action_state_to_metastate.go

```diff
@@ -0,0 +1,467 @@
+func (act ActionStateToMetaState) getStateToMetastateMap(workitemTypeID uuid.UUID) (map[string]string, error) {
+	wit, err := act.Db.WorkItemTypes().Load(act.Ctx, workitemTypeID)
+	if err != nil {
+		return nil, err
+	}
+	stateList, err := act.getValueListFromEnumField(wit, workitem.SystemState)
+	if err != nil {
+		return nil, err
+	}
+	metastateList, err := act.getValueListFromEnumField(wit, workitem.SystemMetaState)
+	if err != nil {
+		return nil, err
+	}
+	if len(stateList) != len(metastateList) {
+		return nil, errs.Errorf("inconsistent number of states and metatstates in the current work item type (must be equal): %d states != %d metastates", len(stateList), len(metastateList))
+	}
+	stateToMetastateMap := make(map[string]string)
+	for idx := range stateList {
+		thisState, ok := stateList[idx].(string)
+		if !ok {
+			return nil, errs.New("state value in value list is not of type string")
+		}
+		thisMetastate, ok := metastateList[idx].(string)
+		if !ok {
+			return nil, errs.New("metastate value in value list is not of type string")
+		}
+		// only the first state<->metatstate mapping needs to be taken as
+		// per definition. So we're only adding a new entry here if there
+		// is not an entry already existing.
+		if _, ok := stateToMetastateMap[thisState]; !ok {
+			stateToMetastateMap[thisState] = thisMetastate
+		}
+	}
+	return stateToMetastateMap, nil
+}
+
+// getMetastateToStateMap returns the mapping from metastate to state values read from the template.
+func (act ActionStateToMetaState) getMetastateToStateMap(workitemTypeID uuid.UUID) (map[string]string, error) {
+	stateToMetastate, err := act.getStateToMetastateMap(workitemTypeID)
+	if err != nil {
+		return nil, err
+	}
+	metastateToStateMap := make(map[string]string)
+	for state, metastate := range stateToMetastate {
+		// this is important: we only add a new entry if there
+		// is not an entry already existing. Therefore, we're only
+		// using the first mapping, satisfying the req for getting the
+		// first matching state for a metastate.
+		if _, ok := metastateToStateMap[metastate]; !ok {
+			metastateToStateMap[metastate] = state
+		}
+	}
+	return metastateToStateMap, nil
+}
+
+func (act ActionStateToMetaState) addOrUpdateChange(changes change.Set, attributeName string, oldValue interface{}, newValue interface{}) change.Set {
+	for _, change := range changes {
+		if change.AttributeName == attributeName {
+			change.NewValue = newValue
+			return changes
+		}
+	}
+	newChanges := append(changes, change.Change{
+		AttributeName: attributeName,
+		OldValue:      oldValue,
+		NewValue:      newValue,
+	})
+	return newChanges
+}
+
+func (act ActionStateToMetaState) fuseChanges(c1 change.Set, c2 change.Set) *change.Set {
+	for _, change := range c2 {
+		c1 = act.addOrUpdateChange(c1, change.AttributeName, change.OldValue, change.NewValue)
+	}
+	return &c1
+}
+
+// OnChange executes the action rule. It implements rules.Action.
```

Comment:
"executes the action rule" doesn't sound right and it is not helpful.

Reply:
The same review comment was made in https://github.com/fabric8-services/fabric8-wit/pull/2249#discussion_r211577820 , which should've been fixed.